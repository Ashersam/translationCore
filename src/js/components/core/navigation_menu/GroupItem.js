// SubMenuItem.js//
//api imports
import React from 'react'
import {Glyphicon} from 'react-bootstrap'
import style from './Style'

class GroupItem extends React.Component {
  render() {
    let active = this.props.contextId == this.props.contextIdReducer.contextId
    let {selections} = this.props.selectionsReducer
    let itemStyle = active ? style.activeSubMenuItem : style.subMenuItem;
    let {reference} = this.props.contextId
    let statusGlyph = <div />
    if (active && selections.length > 0) {
      statusGlyph = <Glyphicon glyph="ok" style={style.menuItem.statusIcon.correct} />
    }
    return (
      <div onClick={() => this.props.actions.changeCurrentContextId(this.props.contextId)}
          style={itemStyle}
          title="Click to select this check">
          {statusGlyph}
          {" " + this.props.resourcesReducer.book + " " + reference.chapter + ":" + reference.verse}
      </div>
    );
  }
}

module.exports = GroupItem;
