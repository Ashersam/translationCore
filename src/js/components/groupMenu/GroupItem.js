import React from 'react'
import ReactDOM from 'react-dom';
import { Glyphicon } from 'react-bootstrap'
import style from './Style';
const MENU_BAR_HEIGHT = 30;
const MENU_ITEM_HEIGHT = 38;

class GroupItem extends React.Component {
  /**
   * @description Generate the proper glyphicon based on selections
   * @return {component} statusGlyph - component to render
   */
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
  }

  componentDidMount() {
    if (this.props.active) {
      if (this.inView(this.props.groupMenuHeader, this)) {
        //If the menu and current check are able to be rendered in the 
        //same window scroll to the group menu item
        this.props.scrollIntoView(this.props.groupMenuHeader);
      }
      else {
        //Scroll to the current check item
        this.props.scrollIntoView(this);
      }
    }
  }

  componentWillReceiveProps(nextProps, context) {
    if (nextProps.active) {
      if (this.inView(nextProps.groupMenuHeader, this)) {
        //If the menu and current check are able to be rendered in the 
        //same window scroll to the group menu item
        nextProps.scrollIntoView(nextProps.groupMenuHeader);
      }
      else {
        //Scroll to the current check item
        nextProps.scrollIntoView(this);
      }
    }
  }

  /**
   * @description - Tests if the the two elements are in the scope of the window (scroll bar)
   * The consts MENU_BAR_HEIGHT & MENU_ITEM_HEIGHT are set to account for the static window avialablity 
   * @param {object} groupMenu - The current group menu header that is extended/actived (i.e. Metaphors)
   * @param {object} currentItem - The current group check item that is active (i.e. Luke 1:1)
   */
  inView(groupMenu, currentItem) {
    var rectGroup = ReactDOM.findDOMNode(groupMenu).getBoundingClientRect();
    var rectItem = ReactDOM.findDOMNode(currentItem).getBoundingClientRect();
    var viewHeight = Math.max(document.documentElement.clientHeight, window.innerHeight);
    return Math.abs(rectGroup.top - rectItem.top) + MENU_BAR_HEIGHT + MENU_ITEM_HEIGHT <= viewHeight;
  }

  onClick() {
    this.props.actions.changeCurrentContextId(this.props.contextId);
  }

  render() {
    let { reference } = this.props.contextId;
    return (
      <div onClick={this.onClick}
        style={this.props.active ? style.activeSubMenuItem : style.subMenuItem}
        title="Click to select this check">
        {this.props.statusGlyph}
        {" " + this.props.bookName + " " + reference.chapter + ":" + reference.verse + this.props.selectionText}
      </div>
    );
  }
}

module.exports = GroupItem;
