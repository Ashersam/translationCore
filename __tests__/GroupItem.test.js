/* eslint-disable no-console */
/* eslint-env jest */
import React from 'react';
import GroupItem from '../src/js/components/groupMenu/GroupItem';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import { Glyphicon } from 'react-bootstrap';

describe('Tests for GroupItem', () => {
  test('Test GroupItem', () => {
    const props = {
      bookName: 'Titus',
      selectionText: 'Paul',
      contextId: {
        reference: {
          chapterVerseMenu: true,
          text: 'Tit 1:1',
          verse: 'Paul'
        }
      },
      actions: {
          changeCurrentContextId: jest.fn()
      },
      statusBadge: <Glyphicon glyph="ok"/>,
      scrollIntoView: jest.fn(),
      inView: jest.fn(),
      active: true,
    };

    // when
    const wrapper = shallow(
      <GroupItem {...props} />
    );

    // then
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
