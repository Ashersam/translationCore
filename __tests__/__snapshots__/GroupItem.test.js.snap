// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for GroupItem Test GroupItem 1`] = `
<div
  className="group-item active"
  onClick={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "var(--accent-color)",
      "borderBottom": "1px solid var(--background-color-dark)",
      "color": "var(--reverse-color)",
      "cursor": "pointer",
      "display": "flex",
      "height": 38,
      "padding": "10px 0",
      "zIndex": 1,
    }
  }
>
  <Glyphicon
    bsClass="glyphicon"
    glyph="ok"
  />
  <span
    className="selection"
    data-class="selection-tooltip"
    data-delay-hide="100"
    data-effect="float"
    data-place="bottom"
    data-tip="Paul"
    data-type="dark"
  >
    <span
      style={
        Object {
          "display": "block",
          "overflow": "hidden",
          "padding": "0px 20px 0px 0px",
          "textOverflow": "ellipsis",
          "whiteSpace": "nowrap",
        }
      }
    >
      Tit 1:1 Paul
    </span>
  </span>
  <ReactTooltip
    insecure={true}
    resizeHide={true}
    wrapper="div"
  />
</div>
`;
